### ISSUE TITLE
Unused local variable declarations seem not to be considered dependencies

### ISSUE THREAD

**Issue by AukeHaanstra** [TNG/ArchUnit#1137]
Unused local variable declarations or unused objects seem not to be considered dependencies, neither are their corresponding import statements. Is this a limitation of ArchUnit, a bug, or am I missing something?
Only in case of an exception, there is a functional dependency. For me this became apparent when trying to modularize an application whose boundaries were set by ArchUnit. I needed to refactor the code that was catching a dependency which it ought not to have depended on, which for me was rather unexpected.

Please see:
https://github.com/AukeHaanstra/ArchUnitIssue/blob/main/src/test/java/org/example/DependencyAssertionsTest.java

**Comment by hankem** [comment]
Thanks for raising the issue and providing the example project; that's very helpful! 💚

Some of the presumably missing violations are due to the fact that local variables are currently not analyzed (#768) and import "statements" do not even exist at byte code level. Therefore your classes [`F`](https://github.com/AukeHaanstra/ArchUnitIssue/blob/main/src/main/java/org/example/f/F.java) and [`H`](https://github.com/AukeHaanstra/ArchUnitIssue/blob/main/src/main/java/org/example/h/H.java) don't have dependencies to `A` nor `Ex`:
```java
class F {
    void doF() {
        A a = null; // dependency NOT detected
    }
}

class H {
    public void doH() {
        new B().giveEx(); // dependency NOT detected
        Ex ex = new B().giveEx(); // dependency NOT detected
    }
}
```
ArchUnit does not detect a dependency if there is no specific reference in the byte code. Similarly to #1012, this in your class [`D`](https://github.com/AukeHaanstra/ArchUnitIssue/blob/main/src/main/java/org/example/d/D.java):
```java
log.info("Exception from ex", ex);
```
is just an `invokeinterface` of `InterfaceMethod org/slf4j/Logger.info:(Ljava/lang/String;Ljava/lang/Throwable;)V`, but no specific dependency to `Ex`.

**Comment by hankem** [comment]
But I'm actually unsure whether
```java
    public void doD() {
        B b = new B();
        try {
            b.doB();
        } catch (Ex ex) { // dependency NOT detected
            log.info("Exception from ex", ex);
        }
    }
```
shouldn't actually have the dependency to `Ex` that you expected, since `classes.get(D.class).getMethod("doD").getTryCatchBlocks().iterator().next().getCaughtThrowables()` indeed contains `Ex`. 🤔

**Comment by AukeHaanstra** [comment]
So the situation is that ArchUnit could be able to detect that type of dependency, though at this moment, this has not been implemented yet?

**Comment by hankem** [comment]
Yes, exactly: The domain object for the [`TryCatchBlock`](https://www.javadoc.io/doc/com.tngtech.archunit/archunit/latest/com/tngtech/archunit/core/domain/TryCatchBlock.html) is already available, but this information was not added to the [`JavaClassDependencies`](https://github.com/TNG/ArchUnit/blob/v1.0.1/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClassDependencies.java#L44-L53).

**Comment by hankem** [comment]
As a workaround, you can also use a user-defined condition like this:

<details><summary>static imports</summary>

```java
import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideInAPackage;
import static com.tngtech.archunit.lang.SimpleConditionEvent.satisfied;
import static com.tngtech.archunit.lang.conditions.ArchConditions.dependOnClassesThat;
```

</details>

```java
ArchCondition<JavaClass> dependOnClassesInPackage(String packageIdentifier) {
    return dependOnClassesThat(resideInAPackage(packageIdentifier))
        .or(new ArchCondition<JavaClass>("<overridden>") {
            @Override
            public void check(JavaClass javaClass, ConditionEvents events) {
                for (JavaCodeUnit codeUnit : javaClass.getCodeUnits()) {
                    for (TryCatchBlock tryCatchBlock : codeUnit.getTryCatchBlocks()) {
                        tryCatchBlock.getCaughtThrowables().stream()
                            .filter(resideInAPackage(packageIdentifier))
                            .map(throwable -> satisfied(
                                codeUnit,
                                codeUnit.getDescription() + " catches " + throwable.getName() + " in " + tryCatchBlock.getSourceCodeLocation()
                            ))
                            .forEach(events::add);
                    }
                }
            }
        })
        .as("depend on classes in package '%s'", packageIdentifier);
}
```

**Comment by AukeHaanstra** [comment]
Ah, that's a neat workaround, thanks a lot! I updated my example tests accordingly.

**Comment by codecholeric** [comment]
I guess it would make sense to add the types found within this custom condition to be just part of `JavaClassDependencies`, right?

**Comment by plazer1** [comment]
> Yes, exactly: The domain object for the [TryCatchBlock](https://www.javadoc.io/doc/com.tngtech.archunit/archunit/latest/com/tngtech/archunit/core/domain/TryCatchBlock.html) is already available, but this information was not added to the [JavaClassDependencies](https://github.com/TNG/ArchUnit/blob/v1.0.1/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClassDependencies.java#L44-L53).

We need this functionality for reliable extraction of dependencies from bytecode. Please, implement this. Preferably we would like the types of caught exceptions to be present in `JavaCodeUnit.getAccessesFromSelf()` or otherwise easily accessible from `JavaCodeUnit`.

It also appears that types used in `instanceof` and cast expressions are not included in `JavaCodeUnit.getAccessesFromSelf()`:
```
expression instanceof SomeClass
(SomeClass) obj
```
It would be great if these were reported as well.

Another thing that seems to be missing is array creation: `new SomeClass[]` or `SomeClass[]::new`.

**Comment by plazer1** [comment]
Another thing that seems to be missing is array creation: `new SomeClass[]` or `SomeClass[]::new`.
